sql/catalyst/src/main/java/org/apache/spark/sql/connector/distributions/ClusteredDistribution.java

```diff
@@ -32,4 +32,13 @@
    * Returns clustering expressions.
    */
   Expression[] clustering();
+
+  /**
+   * Returns the number of partitions required by this write.
+   * <p>
+   * Implementations may want to override this if it requires the specific number of partitions.
+   *
+   * @return the required number of partitions, non-positive values mean no requirement.
+   */
+  default int requiredNumPartitions() { return 0; }
```

Comment:
Why not also add a parallelism requirement to `UnspecifiedDistribution`? That could insert a coalesce.

Reply:
I'm trying to not over-engineering here; this PR addresses the actual use case and no more. If we'd like to address more, I'd like to see the actual use case or at least possible scenario for each one. Like describing the characteristic of specific storage and how the functionality will help or why the functionality is required.

Reply:
@HeartSaVioR, I think this is part of the use case. If this interface can require some number of incoming partitions, then it should do that in all cases. It makes the behavior of this interface harder to understand if this is not enforced for certain values returned by the other methods.

In addition, this interface is less useful if it can't be used to control the incoming parallelism without also changing the data. I think a coalesce makes a lot more sense than ignoring this.

Reply:
Yeah I thought this a bit more, and agree restricting the "parallelism" is a valid use case regardless of distribution.

"coalesce" vs "repartition" remains the question; when distribution is specified I expect "repartition" will take effect, so probably "repartition" is more consistent. And what I observed from troubleshooting is that coalesce could bring an edge-case if used blindly, like when sink places with the same stage shuffle was done and takes whole stage quite slow.
(That was `coalesce(1)` and we simply fixed the problem via changing it to `repartition(1)`.)

I'd feel safer if writer requirement just affects the writer operation and doesn't affect other operations. What do you think?